In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
pd.set_option('display.max_rows', 15)

In [2]:
df = pd.read_csv("./csv_2.0.csv")
df

Unnamed: 0  firm       q  booktomarketratio  salesannualgrowth  \
0               0   AAP  2000q1                NaN                NaN   
1               1   AAP  2000q2                NaN                NaN   
2               2   AAP  2000q3                NaN                NaN   
3               3   AAP  2000q4                NaN                NaN   
4               4   AAP  2001q1                NaN                NaN   
...           ...   ...     ...                ...                ...   
20785       20785  ZEUS  2021q2              29.39           1.239573   
20786       20786  ZEUS  2021q3              24.36           1.228807   
20787       20787  ZEUS  2021q4              23.50           0.883854   
20788       20788  ZEUS  2022q1              38.46           0.503556   
20789       20789  ZEUS  2022q2              25.75           0.275320   

       operatingprofitgrowth  market_value  
0                        NaN           NaN  
1                        NaN           NaN  
2                        NaN           NaN  
3                        NaN           NaN  
4                        NaN           NaN  
...                      ...           ...  
20785              -5.669606   10398769.80  
20786            -171.239669    9712697.40  
20787               6.756052    9974316.50  
20788               1.681608   17798749.56  
20789               1.269293   12863644.25  

[20790 rows x 7 columns]

In [3]:
df = df.groupby('firm').tail(1)
df

Unnamed: 0  firm       q  booktomarketratio  salesannualgrowth  \
89             89   AAP  2022q2             173.09                NaN   
179           179   ABG  2022q2             169.34           0.528715   
269           269   ACI  2022q2              26.72                NaN   
359           359   AEO  2022q2              11.18           0.019740   
449           449   AKA  2022q2               2.76                NaN   
...           ...   ...     ...                ...                ...   
20429       20429  WEYS  2022q2              24.45           0.291762   
20519       20519  WLFC  2022q2              37.48           0.174555   
20609       20609   WSO  2022q2             238.82           0.153606   
20699       20699  WSTG  2022q2              33.24          -0.099363   
20789       20789  ZEUS  2022q2              25.75           0.275320   

       operatingprofitgrowth  market_value  
89                       NaN  5.036365e+08  
179                 1.536172  4.081771e+08  
269                      NaN  1.087664e+08  
359                 0.314044  1.546201e+07  
449                      NaN  1.177259e+06  
...                      ...           ...  
20429               1.247248  4.974035e+06  
20519               1.613103  1.405103e+07  
20609               1.324352  5.301267e+08  
20699               1.681130  1.808688e+06  
20789               1.269293  1.286364e+07  

[231 rows x 7 columns]

In [4]:
df = df.dropna()
df.head(10)

Unnamed: 0  firm       q  booktomarketratio  salesannualgrowth  \
179          179   ABG  2022q2             169.34           0.528715   
359          359   AEO  2022q2              11.18           0.019740   
539          539  AMZN  2022q2             106.21           0.072108   
629          629    AN  2022q2             111.76          -0.015648   
719          719   ANF  2022q2              16.92           0.040129   
809          809  APRN  2022q2               3.64           0.001830   
899          899   ASO  2022q2              35.54          -0.071253   
989          989   AZO  2022q2            2149.12           0.058668   
1079        1079  BBBY  2022q2               4.97          -0.250993   
1169        1169   BBW  2022q2              16.42           0.283217   

      operatingprofitgrowth  market_value  
179                1.536172  4.081771e+08  
359                0.314044  1.546201e+07  
539                0.430667  1.395621e+10  
629                1.052622  2.598420e+08  
719               -0.169345  1.192530e+07  
809                1.842422  1.907287e+05  
899                0.859537  5.432317e+07  
989                0.977803 -7.279564e+09  
1079               4.718917 -1.094881e+06  
1169               1.380858  1.618010e+06

In [5]:
data = pd.DataFrame(df[['booktomarketratio', 'salesannualgrowth', 'market_value' ,'operatingprofitgrowth']])
data

booktomarketratio  salesannualgrowth  market_value  \
179               169.34           0.528715  4.081771e+08   
359                11.18           0.019740  1.546201e+07   
539               106.21           0.072108  1.395621e+10   
629               111.76          -0.015648  2.598420e+08   
719                16.92           0.040129  1.192530e+07   
...                  ...                ...           ...   
20429              24.45           0.291762  4.974035e+06   
20519              37.48           0.174555  1.405103e+07   
20609             238.82           0.153606  5.301267e+08   
20699              33.24          -0.099363  1.808688e+06   
20789              25.75           0.275320  1.286364e+07   

       operatingprofitgrowth  
179                 1.536172  
359                 0.314044  
539                 0.430667  
629                 1.052622  
719                -0.169345  
...                      ...  
20429               1.247248  
20519               1.613103  
20609               1.324352  
20699               1.681130  
20789               1.269293  

[209 rows x 4 columns]

In [6]:
def dotProduct(mtx1, mtx2):
   return mtx1.dot(mtx2)

In [7]:
def cov(matrix):
    matrix_full_ones = np.full((matrix.shape[0], 1), 1)
    y_hat = (dotProduct(matrix.T, matrix) -(1/matrix.shape[0])*( dotProduct(matrix.T,matrix_full_ones)).dot(dotProduct(matrix.T, matrix_full_ones).T))
    result = (1/(matrix.shape[0]-1))
    return (result * y_hat)

In [8]:

def corr(matrix):
    Dinv = np.diag(1 / np.sqrt(np.diag(matrix)))
    corr = Dinv @ matrix @ Dinv
    return corr

In [9]:
corr(cov(data))

0         1         2         3
0  1.000000 -0.001943  0.446868  0.030576
1 -0.001943  1.000000  0.002012  0.099941
2  0.446868  0.002012  1.000000  0.013911
3  0.030576  0.099941  0.013911  1.000000

Based on the table shown above, we know that if the value of the correlation matrix is -1, this indicates a perfectly negative linear correlation between this 2 variables.
If the value we have is 0, it indicates no linear correlation between this two variables.
If the value we have is 1, it indicates a perfectly positive linear correlation between this two variables.

We can see that the only perfect correlations are between the values with themselves, which has logic. 

Also we can see that the variables that have a high correlation are booktomarketratio with market_value	with a correlation value of 0.446868. 

In [10]:
data.corr()

booktomarketratio  salesannualgrowth  market_value  \
booktomarketratio               1.000000          -0.001943      0.446868   
salesannualgrowth              -0.001943           1.000000      0.002012   
market_value                    0.446868           0.002012      1.000000   
operatingprofitgrowth           0.030576           0.099941      0.013911   

                       operatingprofitgrowth  
booktomarketratio                   0.030576  
salesannualgrowth                   0.099941  
market_value                        0.013911  
operatingprofitgrowth               1.000000

In [11]:
cov_matrix = data.cov()
cov_matrix

booktomarketratio  salesannualgrowth  market_value  \
booktomarketratio           9.854911e+04      -2.910168e-01  2.791045e+11   
salesannualgrowth          -2.910168e-01       2.275845e-01  1.909290e+06   
market_value                2.791045e+11       1.909290e+06  3.958420e+18   
operatingprofitgrowth       6.494736e+01       3.225999e-01  1.872698e+08   

                       operatingprofitgrowth  
booktomarketratio               6.494736e+01  
salesannualgrowth               3.225999e-01  
market_value                    1.872698e+08  
operatingprofitgrowth           4.578283e+01

*Variance:* 

The variance is the measure of how far the individual values in a data are from the mean value. 

Is used to quantify spread or dispersion that describes how much variability there is in it.

If we have a high variance, this tells us that the values in our data are far from their mean, having a high variability.

If we have a low variance, this tells us that the values are close to the mean, having a low variability.

Covariance:

The covariance is a measure of the strength of the correlation between two or more sets of random variates. It evaluates how much to what extent the variables change together. 

If we have a positive covariance between 2 variables this means that this 2 variables tend to move in the same direction.

If we have a negative covariance between 2 variables this means that this 2 variables tend to move in inverse directions.

Correlation:

The correlation is a measure that expresses the extent to which two variables are linearly related, this means they change together at a constant rate.

It is measured by the coefficient "r" that is a value between -1 and +1 that quantifies the strength of the relationship between 2 variables.

The closer the value is to 1 or -1, the stronger the relationship. This means that as one value increases, so does the other one.

The closer the value is to 0, the weaker the relationship. This means that as one value increases, the other decreases.



In [12]:
matrix = data.corr()
matrix

booktomarketratio  salesannualgrowth  market_value  \
booktomarketratio               1.000000          -0.001943      0.446868   
salesannualgrowth              -0.001943           1.000000      0.002012   
market_value                    0.446868           0.002012      1.000000   
operatingprofitgrowth           0.030576           0.099941      0.013911   

                       operatingprofitgrowth  
booktomarketratio                   0.030576  
salesannualgrowth                   0.099941  
market_value                        0.013911  
operatingprofitgrowth               1.000000

In [13]:
def getL(l):
    sum = l.sum()
    shape = l.shape[0]
    return sum/shape

In [14]:
def leveragePoints(mtx):
    leverages = np.diagonal(mtx.dot(np.linalg.inv(mtx.T .dot(mtx)).dot(mtx.T)))
    boolean_mtx = leverages > (3 * getL(leverages))
    real_values_of_mtx = np.nonzero(boolean_mtx)
    return real_values_of_mtx[0]

In [15]:
lp  = pd.DataFrame(leveragePoints(data))
lp

0
0    2
1    7
2   27
3   34
4   36
5  132
6  167
7  169
8  178
9  190

In [16]:
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn import preprocessing

def calculate_vif(df, features):    
    df = df.dropna()
    vif, tolerance = {}, {} 
    for feature in features:
        
        X = [f for f in features if f != feature]        
        X, y = df[X], df[feature]  
        r2 = LinearRegression().fit(X, y).score(X, y)                
        
        
        tolerance[feature] = 1 - r2
        vif[feature] = 1/(tolerance[feature])
    return pd.DataFrame({'VIF': vif, 'Tolerance': tolerance})

Multicollinearity analysis:

Multicollinearity is the strong linear dependence relationship between more than 2 groups of explanatory variables in a multiple regression.

In other words, it is the high correlation between two or more explanatory variables.

The greater the value is closer to 1, this means that the variables are highly correlated.

In [17]:
X = data[['booktomarketratio','salesannualgrowth', 'operatingprofitgrowth']]
calculate_vif(df=X, features=['booktomarketratio','salesannualgrowth', 'operatingprofitgrowth'])

VIF  Tolerance
booktomarketratio      1.000961   0.999040
salesannualgrowth      1.010114   0.989987
operatingprofitgrowth  1.011056   0.989065

In [18]:
import statsmodels.api as sm


# storing dependant values
Y = data['market_value']

# storing independent values
X = data[['booktomarketratio','salesannualgrowth', 'operatingprofitgrowth']]


X = sm.add_constant(X)

# fit the model
model = sm.OLS(Y, X, missing="drop").fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:           market_value   R-squared:                       0.200
Model:                            OLS   Adj. R-squared:                  0.188
Method:                 Least Squares   F-statistic:                     17.05
Date:                Fri, 14 Oct 2022   Prob (F-statistic):           6.33e-10
Time:                        20:47:11   Log-Likelihood:                -4747.7
No. Observations:                 209   AIC:                             9503.
Df Residuals:                     205   BIC:                             9517.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                  8.183e+